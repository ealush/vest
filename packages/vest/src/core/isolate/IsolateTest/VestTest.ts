import { invariant, isPromise, optionalFunctionValue } from 'vest-utils';

import type { TIsolateTest } from 'IsolateTest';
import {
  TestAction,
  TestStateMachineAction,
  TestStatus,
  createTestStateMachine,
} from 'IsolateTestStateMachine';
import { TestSeverity } from 'Severity';
import { TFieldName, TGroupName } from 'SuiteResultTypes';

const TestStateMachine = createTestStateMachine();

export class VestTest {
  // Read

  static getData<
    F extends TFieldName = TFieldName,
    G extends TGroupName = TGroupName
  >(test: TIsolateTest<F, G>) {
    invariant(test.data);
    return test.data;
  }

  static warns(test: TIsolateTest): boolean {
    return VestTest.getData(test).severity === TestSeverity.Warning;
  }

  static isPending(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PENDING);
  }

  static isOmitted(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.OMITTED);
  }

  static isUntested(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.UNTESTED);
  }

  static isFailing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.FAILED);
  }

  static isCanceled(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.CANCELED);
  }

  static isSkipped(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.SKIPPED);
  }

  static isPassing(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.PASSING);
  }

  static isWarning(test: TIsolateTest): boolean {
    return VestTest.statusEquals(test, TestStatus.WARNING);
  }

  static hasFailures(test: TIsolateTest): boolean {
    return VestTest.isFailing(test) || VestTest.isWarning(test);
  }

  static isNonActionable(test: TIsolateTest): boolean {
    return (
      VestTest.isSkipped(test) ||
      VestTest.isOmitted(test) ||
      VestTest.isCanceled(test)
    );
  }

  static isTested(test: TIsolateTest): boolean {
    return VestTest.hasFailures(test) || VestTest.isPassing(test);
  }

  static awaitsResolution(test: TIsolateTest): boolean {
    // Is the test in a state where it can still be run, or complete running
    // and its final status is indeterminate?
    return (
      VestTest.isSkipped(test) ||
      VestTest.isUntested(test) ||
      VestTest.isPending(test)
    );
  }

  static isAsyncTest(test: TIsolateTest): boolean {
    return isPromise(VestTest.getData(test).asyncTest);
  }

  static statusEquals(test: TIsolateTest, status: TestStatus): boolean {
    return VestTest.getData(test).status === status;
  }

  // Mutate

  static setPending(test: TIsolateTest) {
    VestTest.setStatus(test, TestStatus.PENDING);
  }

  static fail(test: TIsolateTest): void {
    VestTest.setStatus(
      test,
      VestTest.warns(test) ? TestStatus.WARNING : TestStatus.FAILED
    );
  }

  static pass(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.PASSING);
  }

  static warn(test: TIsolateTest): void {
    VestTest.setData(test, current => ({
      ...current,
      severity: TestSeverity.Warning,
    }));
  }

  static setData(
    test: TIsolateTest,
    setter:
      | ((current: TIsolateTest['data']) => TIsolateTest['data'])
      | TIsolateTest['data']
  ): void {
    test.data = optionalFunctionValue(setter, VestTest.getData(test));
  }

  static skip(test: TIsolateTest, force?: boolean): void {
    // Without this force flag, the test will be marked as skipped even if it is pending.
    // This means that it will not be counted in "allIncomplete" and its done callbacks
    // will not be called, or will be called prematurely.
    // What this mostly say is that when we have a pending test for one field, and we then
    // start typing in a different field - the pending test will be canceled, which
    // is usually an unwanted behavior.
    // The only scenario in which we DO want to cancel the async test regardless
    // is when we specifically skip a test with `skipWhen`, which is handled by the
    // "force" boolean flag.
    // I am not a fan of this flag, but it gets the job done.
    VestTest.setStatus(test, TestStatus.SKIPPED, force);
  }

  static cancel(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.CANCELED);
    VestTest.getData(test).abortController.abort(TestStatus.CANCELED);
  }

  static omit(test: TIsolateTest): void {
    VestTest.setStatus(test, TestStatus.OMITTED);
  }

  static reset(test: TIsolateTest): void {
    VestTest.setStatus(test, TestAction.RESET);
  }

  static setStatus(
    test: TIsolateTest,
    status: TestStateMachineAction,
    payload?: any
  ): void {
    VestTest.setData(test, current => ({
      ...current,
      status: TestStateMachine.staticTransition(
        current.status,
        status,
        payload
      ),
    }));
  }
}
