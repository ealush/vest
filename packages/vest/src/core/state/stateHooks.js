import ctx from 'ctx';

const getStateRef = () => ctx.use().stateRef;

export function usePending() {
  return getStateRef().pending();
}
export function useSuiteId() {
  return getStateRef().suiteId();
}
export function useTestCallbacks() {
  return getStateRef().testCallbacks();
}
export function useTestObjects() {
  return getStateRef().testObjects();
}
export function useSkippedTests() {
  return getStateRef().skippedTests();
}
export function useOptionalFields() {
  return getStateRef().optionalFields();
}
