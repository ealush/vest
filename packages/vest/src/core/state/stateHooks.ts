import ctx from 'ctx';

export function useCarryOverTests() {
  return useStateRef().carryOverTests();
}
export function usePending() {
  return useStateRef().pending();
}
export function useLagging() {
  return useStateRef().lagging();
}
export function useSuiteId() {
  return useStateRef().suiteId();
}
export function useTestCallbacks() {
  return useStateRef().testCallbacks();
}
export function useTestObjects() {
  return useStateRef().testObjects();
}
export function useSkippedTests() {
  return useStateRef().skippedTests();
}
export function useOptionalFields() {
  return useStateRef().optionalFields();
}
export function useStateRef() {
  return ctx.use()!.stateRef!;
}
