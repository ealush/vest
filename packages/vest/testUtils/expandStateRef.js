export default function expandStateRef(stateRef) {
  const state = {};
  for (const key in stateRef) {
    const [value] = stateRef[key]();
    state[key] = value;
  }
  return state;
}
