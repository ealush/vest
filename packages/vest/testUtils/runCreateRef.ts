import createState from 'vast';

import createStateRef from 'createStateRef';

export default (
  state?: ReturnType<typeof createState>
): ReturnType<typeof createStateRef> =>
  createStateRef(state ? state : createState(), {
    suiteId: '1000',
  });
