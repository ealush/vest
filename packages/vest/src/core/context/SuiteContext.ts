import { createCascade } from 'context';
import { assign, TinyState, tinyState, cache, CacheApi } from 'vest-utils';
import { DynamicValue } from 'vest-utils/src/utilityTypes';

import { IsolateTest } from 'IsolateTest';
import { Modes } from 'Modes';

export const SuiteContext = createCascade<CTXType>((ctxRef, parentContext) => {
  if (parentContext) {
    return null;
  }

  return assign(
    {
      inclusion: {},
      mode: tinyState.createTinyState<Modes>(Modes.EAGER),
      suiteParams: [],
      testMemoCache,
    },
    ctxRef
  );
});

type CTXType = {
  inclusion: Record<string, DynamicValue<boolean>>;
  mode: TinyState<Modes>;
  suiteParams: any[];
  testMemoCache: CacheApi<IsolateTest>;
  currentTest?: IsolateTest;
  groupName?: string;
  skipped?: boolean;
  omitted?: boolean;
};

export function useCurrentTest(msg?: string) {
  return SuiteContext.useX(msg).currentTest;
}

export function useGroupName() {
  return SuiteContext.useX().groupName;
}

export function useInclusion() {
  return SuiteContext.useX().inclusion;
}

export function useMode() {
  return SuiteContext.useX().mode();
}

export function useSkipped() {
  return SuiteContext.useX().skipped ?? false;
}

export function useOmitted() {
  return SuiteContext.useX().omitted ?? false;
}

const testMemoCache = cache<IsolateTest>(10);

export function useTestMemoCache() {
  return SuiteContext.useX().testMemoCache;
}

export function useSuiteParams() {
  return SuiteContext.useX().suiteParams;
}
