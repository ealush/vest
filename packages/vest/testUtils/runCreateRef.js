import state from 'state';
import usePending from 'usePending';
import useSuiteId from 'useSuiteId';
import useTestCallbacks from 'useTestCallbacks';
import useTestObjects from 'useTestObjects';

export default () =>
  state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [1000, 'suite_name']],
    useTestCallbacks,
    useTestObjects,
  });
