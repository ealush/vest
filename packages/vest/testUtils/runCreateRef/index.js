import useTestCallbacks from '../../src/core/produce/useTestCallbacks';
import state from '../../src/core/state';
import useSuiteId from '../../src/core/suite/useSuiteId';
import usePending from '../../src/core/test/lib/pending/usePending';
import useTestObjects from '../../src/core/test/useTestObjects';

export default () =>
  state.createRef({
    usePending,
    useSuiteId: [useSuiteId, [1000, 'suite_name']],
    useTestCallbacks,
    useTestObjects,
  });
