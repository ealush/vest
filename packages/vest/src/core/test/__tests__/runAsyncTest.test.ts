import _ from 'lodash';

import expandStateRef from '../../../../testUtils/expandStateRef';
import runCreateRef from '../../../../testUtils/runCreateRef';

import VestTest from 'VestTest';
import context from 'ctx';
import { setPending } from 'pending';
import runAsyncTest from 'runAsyncTest';
import { usePending, useTestCallbacks } from 'stateHooks';


const message = 'some message string';

const CASE_PASSING = 'passing';
const CASE_FAILING = 'failing';

let stateRef;
const getCtx = () => ({ stateRef });

describe.each([CASE_PASSING, CASE_FAILING])('runAsyncTest: %s', testCase => {
  let testObject, fieldName;

  const runRunAsyncTest = (testObject: VestTest) =>
    context.run(
      {
        stateRef,
      },
      () => runAsyncTest(testObject)
    );

  beforeEach(() => {
    fieldName = 'field_1';
    stateRef = runCreateRef();
    context.run({ stateRef }, () => {
      const [, setTestCallbacks] = useTestCallbacks();
      setTestCallbacks(state => {
        state.fieldCallbacks = {
          ...state.fieldCallbacks,
          [fieldName]: state.fieldCallbacks[fieldName] || [],
        };

        return state;
      });
    });
    testObject = new VestTest(fieldName, jest.fn(), {
      message,
    });
    testObject.asyncTest =
      testCase === CASE_PASSING ? Promise.resolve() : Promise.reject();
    context.run({ stateRef }, () => {
      setPending(testObject);
    });
  });

  describe('State updates', () => {
    it.withContext(
      'Initial state matches snapshot (sanity)',
      () => {
        const [pending] = usePending();
        expect(pending).toContain(testObject);
        expect(expandStateRef(stateRef)).toMatchSnapshot();
        runRunAsyncTest(testObject);
      },
      getCtx
    );

    it('Should remove test from pending array', () =>
      new Promise<void>(done => {
        runRunAsyncTest(testObject);
        setTimeout(
          context.bind({ stateRef }, () => {
            const [pending] = usePending();
            expect(pending).not.toContain(testObject);
            done();
          })
        );
      }));
  });

  describe('doneCallbacks', () => {
    let fieldCallback_1, fieldCallback_2, doneCallback;
    beforeEach(() => {
      fieldCallback_1 = jest.fn();
      fieldCallback_2 = jest.fn();
      doneCallback = jest.fn();

      context.run({ stateRef }, () => {
        const [, setTestCallbacks] = useTestCallbacks();

        setTestCallbacks(state => ({
          fieldCallbacks: {
            ...state.fieldCallbacks,
            [fieldName]: (state.fieldCallbacks[fieldName] || []).concat(
              fieldCallback_1,
              fieldCallback_2
            ),
          },
          doneCallbacks: state.doneCallbacks.concat(doneCallback),
        }));
      });
    });
    describe('When no remaining tests', () => {
      it('Should run all callbacks', () =>
        new Promise<void>(done => {
          expect(fieldCallback_1).not.toHaveBeenCalled();
          expect(fieldCallback_2).not.toHaveBeenCalled();
          expect(doneCallback).not.toHaveBeenCalled();
          runRunAsyncTest(testObject);
          setTimeout(() => {
            expect(fieldCallback_1).toHaveBeenCalled();
            expect(fieldCallback_2).toHaveBeenCalled();
            expect(doneCallback).toHaveBeenCalled();
            done();
          });
        }));
    });

    describe('When there are more tests left', () => {
      beforeEach(() => {
        context.run({ stateRef }, () => {
          setPending(
            new VestTest('pending_field', jest.fn(), {
              message,
            })
          );
        });
      });

      it("Should only run current field's callbacks", () =>
        new Promise<void>(done => {
          expect(fieldCallback_1).not.toHaveBeenCalled();
          expect(fieldCallback_2).not.toHaveBeenCalled();
          expect(doneCallback).not.toHaveBeenCalled();
          runRunAsyncTest(testObject);
          setTimeout(() => {
            expect(fieldCallback_1).toHaveBeenCalled();
            expect(fieldCallback_2).toHaveBeenCalled();
            expect(doneCallback).not.toHaveBeenCalled();
            done();
          });
        }));
    });

    describe('When test is canceled', () => {
      beforeEach(() => {
        context.run({ stateRef }, () => {
          testObject.cancel();
        });
      });

      it('Should return without running any callback', () =>
        new Promise<void>(done => {
          expect(fieldCallback_1).not.toHaveBeenCalled();
          expect(fieldCallback_2).not.toHaveBeenCalled();
          expect(doneCallback).not.toHaveBeenCalled();
          runRunAsyncTest(testObject);
          setTimeout(() => {
            expect(fieldCallback_1).not.toHaveBeenCalled();
            expect(fieldCallback_2).not.toHaveBeenCalled();
            expect(doneCallback).not.toHaveBeenCalled();
            done();
          });
        }));
    });
  });

  describe('testObject', () => {
    let testObjectCopy;

    beforeEach(() => {
      testObject.fail = jest.fn();
      testObjectCopy = _.cloneDeep(testObject);
    });

    if (testCase === CASE_PASSING) {
      it('Should keep test object unchanged', () =>
        new Promise<void>(done => {
          runRunAsyncTest(testObject);
          setTimeout(() => {
            expect(testObject).toEqual(testObjectCopy);
            done();
          });
        }));

      it('Should return without calling testObject.fail', () =>
        new Promise<void>(done => {
          runRunAsyncTest(testObject);
          setTimeout(() => {
            expect(testObject.fail).not.toHaveBeenCalled();
            done();
          });
        }));
    }

    if (testCase === CASE_FAILING) {
      it('Should call testObject.fail', () =>
        new Promise<void>(done => {
          runRunAsyncTest(testObject);
          setTimeout(() => {
            expect(testObject.fail).toHaveBeenCalled();
            done();
          });
        }));

      describe('When rejecting with a message', () => {
        const rejectionString = 'rejection string';
        beforeEach(() => {
          testObject.asyncTest.catch(Function.prototype);
          testObject.asyncTest = Promise.reject(rejectionString);
        });

        it('Should set test message to rejection string', () =>
          new Promise<void>(done => {
            runRunAsyncTest(testObject);
            setTimeout(() => {
              expect(testObject.message).toBe(rejectionString);
              done();
            });
          }));
      });
    }
  });
});
