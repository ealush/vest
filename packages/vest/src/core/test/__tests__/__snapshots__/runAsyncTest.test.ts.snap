// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runAsyncTest: failing State updates Initial state matches snapshot (sanity) 1`] = `
Object {
  "carryOverTests": Array [],
  "lagging": Array [],
  "optionalFields": Object {},
  "pending": Array [
    VestTest {
      "asyncTest": Promise {},
      "canceled": false,
      "failed": false,
      "fieldName": "field_1",
      "id": "8",
      "isWarning": false,
      "message": "some message string",
      "testFn": [MockFunction],
    },
  ],
  "skippedTests": Array [],
  "suiteId": Object {
    "id": "1000",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {
      "field_1": Array [],
    },
  },
  "testObjects": Array [],
}
`;

exports[`runAsyncTest: passing State updates Initial state matches snapshot (sanity) 1`] = `
Object {
  "carryOverTests": Array [],
  "lagging": Array [],
  "optionalFields": Object {},
  "pending": Array [
    VestTest {
      "asyncTest": Promise {},
      "canceled": false,
      "failed": false,
      "fieldName": "field_1",
      "id": "0",
      "isWarning": false,
      "message": "some message string",
      "testFn": [MockFunction],
    },
  ],
  "skippedTests": Array [],
  "suiteId": Object {
    "id": "1000",
  },
  "testCallbacks": Object {
    "doneCallbacks": Array [],
    "fieldCallbacks": Object {
      "field_1": Array [],
    },
  },
  "testObjects": Array [],
}
`;
